src/schoology.js

```diff
@@ -0,0 +1,381 @@
+      signatureMethod: 'PLAINTEXT',
+      nonceLength: 16,
+      requestToken,
+      accessToken,
+      handleError: this.handleError
+    });
+  }
+
+  async build() {
+    if (!this.accessToken && !this.refreshToken) {
+      await this.getTokensFromUser();
+    }
+    return this;
+  }
+
+  /**
+   * Returns a URL used to initiate the authorization process with schoology and fetch
+   * the authorization code
+   */
+   async getAuthorizationURL(options = {}) {
+    try {
+      const result = await this.oAuth.getRequestTokens('/v1/oauth/request_token');
+      const tokenData = result.response;
+
+      await this.cacheRequestToken(tokenData);
+      
+      return OAuth.makeURL( this.hostedUrl, '/oauth/authorize', {
+        'oauth_token': tokenData.token,
+        'oauth_callback': this.redirectUri,
+      } )
+    } catch ( error ) {
+      this.handleError(error)
+    }
+  }
+
+  /**
+   * Fetches the access and refresh tokens for a valid authorization code
+   */
+   async getAccessTokens() {
+    try {
+      const result = await this.oAuth.getAccessTokens('/v1/oauth/access_token');
+      const tokenData = result.response;
+      
+      await this.setUserAccessToken( tokenData );
+      
+      return tokenData;
+    } catch ( error ) {
+      this.handleError(error)
+    }
+  }
+
+  async getProfile() {
+    try {
+      const resp = await this.makeRequest({
+        url: '/v1/users/me',
+        method: 'GET',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+      });
+      this.schoologyUserId = resp.data.id;
+      return resp.data;
+    } catch(err) {
+      throw new LMSError('Unable to fetch user profile', 'schoology.USER_PROFILE_ERROR', {
+        userId: this.userId
+      });
+    }
+  }
+
+  async getTokensFromUser() {
+    try {
+      const { accessToken, refreshToken, info } = await this.getUserToken(this.userId);
+      this.accessToken = accessToken;
+      this.refreshToken = refreshToken;
+      this.schoologyUserId = info.id;
+    } catch (err) {
+      throw new LMSError('Unable to fetch tokens from user', 'schoology.TOKEN_FETCH_ERROR', {
+        userId: this.userId,
+        message: err.message,
+      });
+    }
+  }
+
+  /**
+   * Handles some schoology API errors
+   */
+  handleError(err, code) {
+    if (err.response) {
+      switch (err.response.status) {
+        case 400:
+          if (err.error === 'invalid_grant') {
+            throw new LMSError('Invalid authorization code', 'schoology.INVALID_AUTH_CODE', {
+              message: err.message,
+              code,
+            });
+          }
+          break;
+        default:
+          throw new LMSError('An error occured', 'schoology.UKW', {
+            message: err.message,
+            stack: err.stack,
+          });
+      }
+    } else {
+      throw new LMSError('An error occured', 'schoology.UKW', {
+        message: err.message,
+      });
+    }
+  }
+
+  isTokenExpired(err) {
```

Comment:
I don't think this would apply for schoology, this was a canvas-specific condition